### PR TITLE
[plugin.video.catchuptvandmore@matrix] 0.2.38+matrix.1

### PR DESCRIPTION
**Pull request by joaopa00**
- [x] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-plugins/blob/master/CONTRIBUTING.md) document
- [x] Each add-on submission should be a single commit with using the following style: [plugin.video.foo] v1.0.0

Once more, this PR is pretty huge. If this PR is merged very quickly, I will send the newest version very soon. So, the PR for the next coming version 0.2.39 should be relatively small, so easier to review.

Thanks for your understanding.

### CODE REVIEW

File: plugin.video.catchuptvandmore/resources/lib/openvpn.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
+# /*
+# *
+# * OpenVPN for Kodi.
```
Comment (enen92) [inline plugin.video.catchuptvandmore/resources/lib/openvpn.py]: I highly doubt we want this on our repository.....

**Comment by joaopa00** [comment]
I removed the problematic feature. Thank you for the review and the feedback.

(Not so) unrelated note: I pushed some logos and picture in the repository repo-resources. Can you merge them?

File: plugin.video.catchuptvandmore/resources/lib/channels/sk/joj.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
+# -*- coding: utf-8 -*-
```
Comment (pkscout) [inline plugin.video.catchuptvandmore/resources/lib/channels/sk/joj.py:1]: the empty __init__.py file above is not needed with Python3. Please remove.

File: plugin.video.catchuptvandmore/resources/lib/channels/tr/beyaz.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
```
Comment (pkscout) [inline plugin.video.catchuptvandmore/resources/lib/channels/tr/beyaz.py:1]: the empty __init__.py file above is not needed with Python3. Please remove.

File: plugin.video.catchuptvandmore/resources/lib/enum/LICENSE [hunk 4]
```diff
@@ -0,0 +1,32 @@
+Copyright (c) 2013, Ethan Furman.
```
Comment (pkscout) [inline plugin.video.catchuptvandmore/resources/lib/enum/LICENSE]: I'm pretty sure that we don't like having Python modules embedded in addons.  The correct process is to create the needed module as an addon and then making it a required dependency.

File: plugin.video.catchuptvandmore/resources/lib/streams/m3u8.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
+# -*- coding: utf-8 -*-
```
Comment (pkscout) [inline plugin.video.catchuptvandmore/resources/lib/streams/m3u8.py:1]: the empty __init__.py file above is not needed with Python3. Please remove.

File: plugin.video.catchuptvandmore/resources/lib/xmltv_utils/sd_json.py [hunk 6]
```diff
@@ -0,0 +1,424 @@
+#!/usr/bin/env python3
```
Comment (pkscout) [inline plugin.video.catchuptvandmore/resources/lib/xmltv_utils/sd_json.py:1]: the empty __init__.py file above is not needed with Python3. Please remove.

**Review by pkscout — CHANGES_REQUESTED** [review]

File: plugin.video.catchuptvandmore/resources/lib/enum/LICENSE (hunk 4 above)
Comment (basrieter) [reply plugin.video.catchuptvandmore/resources/lib/enum/LICENSE]: Nope, we don't. There needs to be a separate submission for this module. There were previous attempts, but no add-on using it. See https://github.com/xbmc/repo-scripts/pull/1334 for the latest. It was for Leia. So there should be new Python3 version submitted to the Matrix repo-scripts.

**Comment by joaopa00** [comment]
Thank you for the feedback. I will send the library for enum to repo-scripts soon.

**Comment by joaopa00** [comment]
I sent a PR to repo-scripts to merge enum. Once that's done, I shall update this PR to take advantage of this feature.

**Comment by joaopa00** [comment]
I removed the dependency on enum. Please, have a look again.

**Comment by joaopa00** [comment]
All the requested changes were done. Is it possible to have a new feedback?
Thanks in advance?

File: plugin.video.catchuptvandmore/resources/lib/channels/sk/joj.py (hunk 2 above)
Comment (joaopa00) [reply plugin.video.catchuptvandmore/resources/lib/channels/sk/joj.py:1]: fixed

File: plugin.video.catchuptvandmore/resources/lib/channels/tr/beyaz.py (hunk 3 above)
Comment (joaopa00) [reply plugin.video.catchuptvandmore/resources/lib/channels/tr/beyaz.py:1]: fixed

**Comment by joaopa00** [comment]
What should I do more to have a review, please?

**Comment by basrieter** [comment]
I will check tonight.

**Comment by joaopa00** [comment]
Thanks

File: plugin.video.catchuptvandmore/resources/lib/channels/ca/tva.py [hunk 7]
```diff
@@ -258,19 +258,15 @@ def get_video_url(plugin,
...
-            'User-Agent':
-            web_utils.get_random_ua(),
-            'Accept':
-            'application/json;pk=%s' %
-            (get_brightcove_policy_key(data_account, data_player)),
-            'X-Forwarded-For':
-            plugin.setting.get_string('header_x-forwarded-for')
+            'User-Agent': web_utils.get_random_ua(),
+            'Accept': 'application/json;pk=%s' % (get_brightcove_policy_key(data_account, data_player)),
+            'X-Forwarded-For': plugin.setting.get_string('header_x-forwarded-for')
```
Comment (basrieter) [inline plugin.video.catchuptvandmore/resources/lib/channels/ca/tva.py]: We don't want add-ons to circumvent GEO blocks, both in streaming and downloading. That means that using `x-forwarded-for` is not allowed.

File: plugin.video.catchuptvandmore/resources/lib/channels/jp/tver.py [hunk 8]
```diff
@@ -17,115 +17,104 @@
...
-            'X-Forwarded-For':
-            plugin.setting.get_string('header_x-forwarded-for')
-        })
-    json_parser = json.loads(resp.text)
+    file_js = urlquick.get(URL_BRIGHTCOVE_POLICY_KEY % (data_account, data_player))
+    policy = re.compile('policyKey:"(.+?)"').findall(file_js.text)[0]
+    headers = {
+        'User-Agent': web_utils.get_random_ua(),
+        'Accept': 'application/json;pk=%s' % policy,
+        'X-Forwarded-For': plugin.setting.get_string('header_x-forwarded-for')
```
Comment (basrieter) [inline plugin.video.catchuptvandmore/resources/lib/channels/jp/tver.py]: We don't want add-ons to circumvent GEO blocks, both in streaming and downloading. That means that using `x-forwarded-for` is not allowed.

File: plugin.video.catchuptvandmore/resources/lib/resolver_proxy.py [hunk 9]
```diff
@@ -200,28 +318,36 @@ def get_brightcove_video_json(plugin,
...
-            'application/json;pk=%s' %
-            (get_brightcove_policy_key(data_account, data_player)),
-            'X-Forwarded-For':
-            plugin.setting.get_string('header_x-forwarded-for')
-        })
-    json_parser = json.loads(resp.text)
+    headers = {
+        'User-Agent': web_utils.get_random_ua(),
+        'Accept': 'application/json;pk=%s' % key,
+        'X-Forwarded-For': plugin.setting.get_string('header_x-forwarded-for')
```
Comment (basrieter) [inline plugin.video.catchuptvandmore/resources/lib/resolver_proxy.py]: We don't want add-ons to circumvent GEO blocks, both in streaming and downloading. That means that using `x-forwarded-for` is not allowed.

File: plugin.video.catchuptvandmore/resources/lib/resolver_proxy.py [hunk 10]
```diff
@@ -300,7 +423,8 @@ def get_francetv_video_stream(plugin,
         final_video_url = json_parser2['url']
         if download_mode:
             return download.download_video(final_video_url)
-        return final_video_url + '|X-Forwarded-For=' + '2.' + str(randint(0, 15)) + '.' + str(randint(0, 255)) + '.' + str(randint(0, 255)) + '&User-Agent=' + web_utils.get_random_ua()
+        return final_video_url + '|X-Forwarded-For=' + '2.' + str(randint(0, 15)) + '.' + str(
```
Comment (basrieter) [inline plugin.video.catchuptvandmore/resources/lib/resolver_proxy.py]: We don't want add-ons to circumvent GEO blocks, both in streaming and downloading. That means that using `x-forwarded-for` is not allowed.

**Review by basrieter — CHANGES_REQUESTED** [review]
We don't want add-ons to circumvent GEO blocks, both in streaming and downloading. That means that using `x-forwarded-for` is not allowed. With Kodi can should only be allowed to stream what can be watched via the official websites.

It seems that it was there before, but it **should not** have been. Please remove the code and settings related to this.

File: plugin.video.catchuptvandmore/resources/lib/resolver_proxy.py (hunk 10 above)
Comment (joaopa00) [reply plugin.video.catchuptvandmore/resources/lib/resolver_proxy.py]: Thanks for the feedback. I removed the problematic code.

**Review by basrieter — APPROVED** [review]

**Comment by basrieter** [comment]
Thanks. All looking good too me now. @pkscout do you agree?

**Comment by joaopa00** [comment]
Thanks for the review. I proposed a PR in repo/resources that looks staled too. Can you merge it too, please?

**Comment by basrieter** [comment]
> Thanks for the review. I proposed a PR in repo/resources that looks staled too. Can you merge it too, please?

Which one?

**Comment by enen92** [comment]
I merged it already

**Comment by joaopa00** [comment]
Thank you very much to everyone for the feedbacks and merging this PR!!

**Pull request merged**